### ISSUE TITLE
query: re passing variables as template args & indicating a return type is a method argument

### ISSUE THREAD

**Issue by SignpostMarv** [vimeo/psalm#2142]
following on from #2135, I've been looking at templating the ext-dom methods;

https://psalm.dev/r/a7cf569ece

If one calls `(<DOMNode> $foo)::appendChild($bar)`, then ideally `if($bar->parentNode === $foo)` should be flagged as a redundant argument- I'm not really sure how to template this up currently.

The particular example linked also makes me wonder how to template up that a return type is the argument type.

What I'd be thinking as the end result is something like this:

```php
/**
* @template T1 as DOMNode|null
*/
class DOMNode {
    /**
    * @readonly
    *
    * @var T1
    */
    public ? DOMNode $parentNode;

    /**
    * @template T as DOMNode
    *
    * @param T $child
    *
    * @return $child<$this>
    */
    public function appendChild(DOMNode $child) : DOMNode {}
}
```

**Comment by AndrolGenhald** [comment]
I don't think this is ever going to be possible. Templates refer to types, not variables.

**Comment by AndrolGenhald** [comment]
```php
/**
* @var Templated<$a>
*/
$b = $b;

if ($b->parentNode === $a) { // should be a redundant condition if method call does not impact state.
  echo 'foo';
}
```

The only way for Psalm to catch that as a redundant condition would be for Psalm to know that `$b->parentNode` refers to the same object as `$a`, and the only way I can think to do that would be to have the `T1` template refer to the actual variable rather than the type. With `$b = Templated<DOMElement>`, `$a` and `$b->parentNode` are possibly different and there's no way to tell the check is redundant.

**Comment by SignpostMarv** [comment]
@AndrolGenhald  I may have been a little knackered last night- I think what I meant to ask was `DOMNode<$this>`

**Comment by AndrolGenhald** [comment]
I don't see how that changes anything, it's still not possible to use a variable as a template type.

I left this open so I could think about it for a bit, but I'm going to go ahead and close it now. This would take a lot of work to implement and it would only work for immutable objects, so I don't see it being used very much.

If it were allowed on mutable objects [this](https://psalm.dev/r/810c50a4ca) would cause problems, which means whenever a possibly mutating method is called the template type would have to be changed from the specific variable to the type of the variable, which makes the feature almost useless.

**Comment by SignpostMarv** [comment]
I'm mostly referring to `$this` being supported by `@return` so there's vaguely-related precedent.

**Comment by AndrolGenhald** [comment]
Ah, in that case, I don't think they're nearly so closely related as you think. I haven't looked into it much, but I think `@return $this` basically just returns the same type with the same templates, it doesn't actually track that it's the same variable: https://psalm.dev/r/e1acaa7a1b
